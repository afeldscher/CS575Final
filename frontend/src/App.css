.header-row {
  margin-top: 1em;
}

.block {
  border: 1px solid black;
  border-radius: 1em;
  padding: 1em;
  margin: 1em;
}

input[type=text]:not(.browser-default):disabled.block-field {
  color: rgb(0, 0, 0);
}

.mine-btn {
  margin: 1em 0;
}

.hash-textarea {
  height: 4em !important;
  color: rgb(0, 0, 0);
}

.materialize-textarea.hash-textarea {
  height: 4.2em;
  color: rgb(0, 0, 0) !important;
  overflow-y: auto;
}

.materialize-textarea.data-textarea {
  height: 10em;
}

.block.col.s11.m5.l5 {
  min-width: 350px;
  padding-top: 1em;
}

.add-block-textarea-wrapper {
  margin-top: 2em !important;
}

.bad-block {
  background-color: #ffcdd2;
}

.mined-block {
  background-color: #c8e6c9;
}