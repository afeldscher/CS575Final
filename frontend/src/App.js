import React from 'react';
import M from 'materialize-css';
import $ from 'jquery';
import './App.css';
import BlockNode from "./js/BlockNode";
import {solveBlock, uuidv4, hasLeadingZeroes} from './js/utils';

function App() {
    return (
        <div className="App">
            <nav>
                <div className="nav-wrapper light-blue center">
                    <a href="#" className="app-title">Blockchain Simulator</a>
                    <ul id="nav-mobile" className="right hide-on-med-and-down">
                        <a className="modal-trigger" href="#infoModal">
                            <i title="About" className="material-icons">info_outline</i>
                        </a>
                    </ul>
                </div>
            </nav>
            <div className="container">
                <BlockChainElement/>
            </div>
        </div>
    );
}

class BlockChainElement extends React.Component {
    constructor(props) {
        super(props);
        this.state = {blocks: []};
        this.addBlock = this.addBlock.bind(this);
        this.mineBlock = this.mineBlock.bind(this);
        this.recalculateAllHashes = this.recalculateAllHashes.bind(this);
        this.handleTamper = this.handleTamper.bind(this);
        this.propagateChanges = this.propagateChanges.bind(this);
        this.resetAllBlocks = this.resetAllBlocks.bind(this);
    }

    addBlock() {
        const blockData = $('#blockDataTextArea').val();
        if (blockData.length === 0) { //todo: Delete this if this is desirable functionality
            M.toast({html: 'Cannot add an empty block.', classes: 'red'});
            return;
        }
        $('#blockDataTextArea').val("");
        const parent = this.state.blocks[this.state.blocks.length - 1];
        let blockNode = new BlockNode(uuidv4(), blockData, parent);
        blockNode.getHash().then(response => {
            blockNode.hash = response.hash;
            this.setState(state => ({
                blocks: [...state.blocks, blockNode]
            }));
        });
    }

    resetAllBlocks() {
        let rootBlock = this.state.blocks[0];
        rootBlock.getHash().then(response => {
            rootBlock.hash = response.hash;
            rootBlock.mined = hasLeadingZeroes(response.hash, $("#numZeroes").val());
            this.propagateChanges(0, rootBlock);
        });
    }

    mineBlock(id) {
        const block = this.state.blocks[id];
        const zeroes = $("#numZeroes").val();
        const numTries = $('#numTries').val();
        solveBlock(block.parent, block.data, zeroes, numTries).then(response => {
            if (response.solved) {
                block.nonce = response.nonce;
                block.hash = response.hash;
                block.mined = true;
                this.propagateChanges(id, block);
            } else {
                M.toast({html: `Unable to solve block #${id + 1} after ${numTries} attempts`, classes: 'red'});
            }
        });
        M.toast({html: `Mining Block #${id + 1}...`, classes: 'green darken-1'});
    }

    async recalculateAllHashes(id, blockArr) {
        for (let i = id; i < this.state.blocks.length - 1; i++) {
            const parentHash = i === 0 ? '0000000000000000000000000000000000000000000000000000000000000000' : blockArr[i].hash;
            let currentBlock = blockArr[i + 1];
            currentBlock.parent = parentHash;
            const response = await currentBlock.getHash();
            currentBlock.hash = response.hash;
            currentBlock.mined = hasLeadingZeroes(response.hash, $("#numZeroes").val());
            blockArr[i + 1] = currentBlock;
        }
        return blockArr;
    }

    handleTamper(id, newData) {
        const block = this.state.blocks[id];
        block.data = newData;
        block.getHash().then(response => {
            block.hash = response.hash;
            block.mined = hasLeadingZeroes(response.hash, $("#numZeroes").val());
            this.propagateChanges(id, block);
        });
    }

    propagateChanges(id, block) {
        let blockArr = this.state.blocks;
        blockArr[id] = block;
        if (id + 1 === this.state.blocks.length) {
            this.setState(state => ({
                blocks: blockArr,
            }));
        } else {
            this.recalculateAllHashes(id, blockArr).then(newBlocks => {
                this.setState(state => ({
                    blocks: blockArr,
                }));
            });
        }
    }

    render() {
        let blocks = this.state.blocks;
        return (
            <div>
                <HeaderRow resetBlockFunction={this.resetAllBlocks} addBlockFunction={this.addBlock}/>
                <div className="row">
                    {blocks.map((it, idx) => (
                        <Block key={idx} id={idx} guid={it.guid} parent={it.parent} data={it.data} nonce={it.nonce}
                               hash={it.hash} mineFunction={this.mineBlock} mined={it.mined}
                               textareaOnChange={this.handleTamper}/>
                    ))}
                </div>
            </div>
        );
    }
}

class Block extends React.Component {
    constructor(props) {
        super(props);
        this.triggerMine = this.triggerMine.bind(this);
    }

    triggerMine() {
        this.props.mineFunction(this.props.id);
    }

    render() {
        let backgroundClass = "block col s11 m5 l5";
        let statusString = "";
        if (this.props.mined) {
            backgroundClass += " mined-block";
            statusString = "Good";
        } else {
            backgroundClass += " bad-block";
            statusString = "Bad"
        }
        return (
            <div
                className={backgroundClass}>
                <div>
                    <div className="block-header">
                        <span className="block-title">Block #{this.props.id + 1}</span><br/>
                        <span>Status: {statusString}</span>
                    </div>
                    <TextBoxInput id={"guid"} value={this.props.guid} label={"GUID"}/>
                    <DisabledTextAreaInput id={"parent"} value={this.props.parent} label={"Parent"}/>
                    <TextAreaInput blockId={this.props.id} id={"data"} value={this.props.data} label={"Block Data"}
                                   onChangeFct={this.props.textareaOnChange}/>
                    <TextBoxInput id={"nonce"} value={this.props.nonce} label={"Nonce"}/>
                    <DisabledTextAreaInput id={"hash"} label={"Hash"} value={this.props.hash}/>
                    <button className="waves-effect waves-light btn mine-btn green darken-1"
                            onClick={this.triggerMine}>Mine
                    </button>
                </div>
            </div>)
    }
}


function TextBoxInput(props) {
    return (
        <div className="input-field">
            <input className="block-field" type="text" id={props.id} disabled={true} value={props.value}/>
            <label className="active">{props.label}</label>
        </div>);
}

function DisabledTextAreaInput(props) {
    return (
        <div className="input-field">
            <textarea readOnly className="materialize-textarea hash-textarea" id={props.id} value={props.value}/>
            <label className="active">{props.label}</label>
        </div>);
}

class TextAreaInput extends React.Component {
    constructor(props) {
        super(props);
        this.triggerOnChangeFunction = this.triggerOnChangeFunction.bind(this);
        this.id = this.props.id + "-" + this.props.blockId;
    }

    triggerOnChangeFunction() {
        const blockData = $('#' + this.id).val();
        this.props.onChangeFct(this.props.blockId, blockData);
    }

    render() {
        return (
            <div className="input-field">
            <textarea onChange={this.triggerOnChangeFunction} className="materialize-textarea data-textarea"
                      id={this.id}>{this.props.value}</textarea>
                <label className="active">{this.props.label}</label>
            </div>);
    }
}

export class HeaderRow extends React.Component {
    DEFAULT_ZEROES = 4;
    DEFAULT_NUM_TRIES = 10000000;

    render() {
        return (
            <div>
                <div className=" header-row row">
                    <div className="col s4">
                        <a title="Click to add a block to the blockchain"
                           href="#addItemModal" className="green darken-1 waves-effect waves-light btn modal-trigger">
                            Add a block
                        </a>
                    </div>
                    <div className="col s4">
                        <div title="Specifies the difficulty for the hash solving algorithm" className="input-field">
                            <input onChange={this.props.resetBlockFunction} defaultValue={this.DEFAULT_ZEROES} min="1" max="32" id="numZeroes" type="number"/>
                            <label htmlFor="numZeroes">Number of Leading Zeroes</label>
                        </div>
                    </div>
                    <div className="col s4">
                        <div title="The maximum number of nonce values to try" className="input-field">
                            <input defaultValue={this.DEFAULT_NUM_TRIES} min="100" id="numTries" type="number"/>
                            <label htmlFor="numTries">Max Tries Per Mine</label>
                        </div>
                    </div>
                </div>
                <div className="modal" id="addItemModal">
                    <div className="modal-content">
                        <h5>Enter data for new block</h5>
                        <div className="input-field add-block-textarea-wrapper">
                            <textarea placeholder="Enter data here" className="materialize-textarea" id="blockDataTextArea"/>
                        </div>
                    </div>
                    <div className="modal-footer">
                        <a href="#" onClick={this.props.addBlockFunction}
                           className="modal-close waves-effect btn-flat">
                            Add
                        </a>
                    </div>
                </div>
                <div className="modal" id="infoModal">
                    <div className="modal-content">
                        <h4>About this Application</h4>
                        <p>This application is a blockchain simulator that uses a proof-of-work consensus algorithm.
                            The specific hashing function is based on the
                            <a href="https://en.bitcoin.it/wiki/Block_hashing_algorithm"> bitcoin hashing algorithm.</a>
                        </p>
                    </div>
                    <div className="modal-footer">
                        <a href="#" className="modal-close waves-effect btn-flat">Close</a>
                    </div>
                </div>
            </div>
        );
    }
}

export default App;
