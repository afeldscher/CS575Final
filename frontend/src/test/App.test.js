import React from 'react';
import App from '../App';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

describe('Test <App>', () => {
  it('renders without crashing', () => {
    shallow(<App />);
  });

  it('matches snapshot', () => {
    const tree = renderer.create(
        <App />,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});