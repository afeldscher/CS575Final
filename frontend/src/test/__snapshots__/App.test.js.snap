// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <App> matches snapshot 1`] = `
<div
  className="App"
>
  <nav>
    <div
      className="nav-wrapper light-blue center"
    >
      <a
        className="app-title"
        href="#"
      >
        Blockchain Simulator
      </a>
      <ul
        className="right hide-on-med-and-down"
        id="nav-mobile"
      >
        <a
          className="modal-trigger"
          href="#infoModal"
        >
          <i
            className="material-icons"
            title="About"
          >
            info_outline
          </i>
        </a>
      </ul>
    </div>
  </nav>
  <div
    className="container"
  >
    <div>
      <div>
        <div
          className=" header-row row"
        >
          <div
            className="col s4"
          >
            <a
              className="green darken-1 waves-effect waves-light btn modal-trigger"
              href="#addItemModal"
              title="Click to add a block to the blockchain"
            >
              Add a block
            </a>
          </div>
          <div
            className="col s4"
          >
            <div
              className="input-field"
              title="Specifies the difficulty for the hash solving algorithm"
            >
              <input
                defaultValue={4}
                id="numZeroes"
                max="32"
                min="1"
                onChange={[Function]}
                type="number"
              />
              <label
                htmlFor="numZeroes"
              >
                Number of Leading Zeroes
              </label>
            </div>
          </div>
          <div
            className="col s4"
          >
            <div
              className="input-field"
              title="The maximum number of nonce values to try"
            >
              <input
                defaultValue={10000000}
                id="numTries"
                min="100"
                type="number"
              />
              <label
                htmlFor="numTries"
              >
                Max Tries Per Mine
              </label>
            </div>
          </div>
        </div>
        <div
          className="modal"
          id="addItemModal"
        >
          <div
            className="modal-content"
          >
            <h5>
              Enter data for new block
            </h5>
            <div
              className="input-field add-block-textarea-wrapper"
            >
              <div>
                <div
                  className="input-field"
                >
                  <input
                    id="checkId"
                    placeholder="Enter check ID"
                    type="text"
                  />
                  <label
                    htmlFor="checkId"
                  >
                    Check Id
                  </label>
                </div>
                <div
                  className="input-field"
                >
                  <textarea
                    className="materialize-textarea"
                    id="checkItems"
                    placeholder="Enter items on bill"
                  />
                  <label
                    htmlFor="checkItems"
                  >
                    Items
                  </label>
                </div>
                <div
                  className="input-field"
                >
                  <input
                    id="checkCost"
                    placeholder="Enter total bill ($)"
                    step="any"
                    type="number"
                  />
                  <label
                    htmlFor="checkCost"
                  >
                    Total Bill
                  </label>
                </div>
                <div
                  className="input-field"
                >
                  <input
                    id="checkTip"
                    placeholder="Enter tip"
                    step="any"
                    type="number"
                  />
                  <label
                    htmlFor="checkTip"
                  >
                    Tip
                  </label>
                </div>
              </div>
            </div>
          </div>
          <div
            className="modal-footer"
          >
            <a
              className="modal-close waves-effect btn-flat"
              href="#"
              id="addItemModalButton"
              onClick={[Function]}
            >
              Add
            </a>
          </div>
        </div>
        <div
          className="modal"
          id="infoModal"
        >
          <div
            className="modal-content"
          >
            <h4>
              About this Application
            </h4>
            <p>
              This application is a blockchain simulator that uses a proof-of-work consensus algorithm. The specific hashing function is based on the
              <a
                href="https://en.bitcoin.it/wiki/Block_hashing_algorithm"
              >
                 bitcoin hashing algorithm.
              </a>
            </p>
          </div>
          <div
            className="modal-footer"
          >
            <a
              className="modal-close waves-effect btn-flat"
              href="#"
            >
              Close
            </a>
          </div>
        </div>
      </div>
      <div
        className="row"
      />
    </div>
  </div>
</div>
`;
